File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.py
import sys
from collections import deque

input = sys.stdin.readline

N, K = map(int, input().split())
belt = list(map(int, input().split()))

robot = deque([])
up = 0
down = N - 1

broken = 0

turn = 0

while broken < K:
    turn += 1

    # step1
    up -= 1
    down -= 1
    if up < 0:
        up += 2 * N
    if down < 0:
        down += 2 * N
    if robot and robot[0] == down:
        robot.popleft()

    # step2
    for idx, r in enumerate(robot):
        next = (r + 1) % (2 * N)
        if belt[next] > 0 and next not in robot:
            robot[idx] = (robot[idx] + 1) % (2 * N)
            belt[next] -= 1
            if belt[next] == 0:
                broken += 1

    if robot and robot[0] == down:
        robot.popleft()

    # step3
    if belt[up] > 0:
        robot.append(up)
        belt[up] -= 1
        if belt[up] == 0:
            broken += 1

print(turn)

File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.js
let input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

let [N, K] = input[0].split(" ").map(Number);
let belt = input[1].split(" ").map(Number);

let robot = [];
up = 0;
down = N - 1;

broken = 0;
turn = 0;

while (broken < K) {
  turn++;

  up--;
  down--;
  if (up < 0) {
    up += 2 * N;
  }
  if (down < 0) {
    down += 2 * N;
  }
  if (robot && robot[0] == down) {
    robot.shift();
  }

  for (let i = 0; i < robot.length; i++) {
    let r = robot[i];
    let next = (r + 1) % (2 * N);
    if (belt[next] > 0 && !robot.includes(next)) {
      robot[i] = (robot[i] + 1) % (2 * N);
      belt[next] -= 1;
      if (belt[next] == 0) {
        broken++;
      }
    }
  }

  if (robot && robot[0] == down) {
    robot.shift();
  }

  if (belt[up] > 0) {
    robot.push(up);
    belt[up] -= 1;
    if (belt[up] == 0) {
      broken += 1;
    }
  }
}

console.log(turn);
